src/file_watching.jl

```diff
@@ -1,56 +1,153 @@
+    path::String
+    mtime::Float64
+    WatchedFile(path::String) = new(path, mtime(path))
+end
+
+"Check if a `WatchedFile` has changed"
+has_changed(wf::WatchedFile) = mtime(wf.path) > wf.mtime
+
+"Set the current state of a `WatchedFile` as unchanged"
+set_unchanged!(wf::WatchedFile) = wf.mtime = mtime(wf.path)
+
 """
-    FileWatcher
+    SimpleWatcher([callback];sleeptime::Float64=0.1)
+
+A simple file watcher. You can specify a callback function, receiving the
+path of each file that has changed as a string, at construction or later
+by the API function described below. The `sleeptime` is the time waited
+between to runs of the loop looking for changed files.
 
-Contains a list of watched files and a callback function that should be applied in the case of file
-events.
+# API functions
+TBD, where `w` is a `SimpleWatcher`. Main API functions (commonly used):
+
+- `start(w)`
+- `stop(w)`
+- `set_callback(w, fcn::Function)`
+- `watch_file(w, path::String)`
+
+Further API functions (probably never used in a normal use case):
+
+- `isrunning(w)`
+- `is_file_watched(w, path::String)`
 """
-struct FileWatcher
-    watched_files::Dict{String,Task}
-    callback::Function
-    FileWatcher(cb::Function) = new(Dict{String,Task}(), cb)
+mutable struct SimpleWatcher
+    callback::Union{Nothing,Function}
+    task::Union{Nothing,Task}
+    sleeptime::Float64
+    filelist::Vector{WatchedFile}
+    SimpleWatcher(fcn::Union{Nothing,Function}=nothing; sleeptime::Float64=0.2) = new(fcn,nothing,max(0.1,sleeptime),Vector{WatchedFile}())
 end
 
 """
-    _file_watcher(filepath, fw)
+    _file_watcher(w::SimpleWatcher)
 
-Helper function that's called asynchronously to act if a file has changed.
-See also [`add_to_filewatcher`](@ref).
+[INTERNAL] Helper function that's spawned as an asynchronous task which
+checks for changed files. Terminates normally upon a `InterruptException`,
+and with a warning for all other exceptions.
 """
-function _file_watcher(filepath::AbstractString, fw::FileWatcher)
-    fileevent = watch_file(filepath, -1) # returns upon a change
-    # if file changed, start new task on this file (s.t. it is still being watched)
-    if !fileevent.renamed && !fileevent.timedout
-        fw.watched_files[filepath] = @async _file_watcher(filepath, fw)
-    else
-        # if renamed or timed out, remove from list of watched files
-        delete!(fw.watched_files, filepath)
+function _file_watcher(w::SimpleWatcher)
+    try
+        while true
+            # checking for changed files
+            changed_files = Vector{String}()
+            foreach(w.filelist) do wf
+                if has_changed(wf)
+                    set_unchanged!(wf)
+                    push!(changed_files, wf.path)
+                end
+            end
+
+            # invoke callback on each changed file (if it's defined), then sleep
+            (w.callback != nothing) && foreach(w.callback, changed_files)
+            sleep(w.sleeptime)
```

Comment:
So instead of building a vector of changed files and then calling on each, can you not do both in one and not have to build the list?

Reply:
This was my first implementation. Then I thought, well, what if the checking for changes takes quite long... and the callback gets set in the middle of it. Then it should execute on the detected changes already. But you're probably right, this is an extremely unlikely case and the code would simplify quite a bit...

So, first check `(w.callback != nothing)`, then enter the checking loop? If there's nothing done upon changes, we don't even have to check.

Reply:
yeah